File: src/detect.ts
import fs from 'fs';
import path from 'path';

export type ProjectLang = 'java' | 'ts' | 'both' | 'unknown';

export interface ProjectInfo {
  lang: ProjectLang;
  root: string;
  name: string;
}

export function detectProject(cwd: string): ProjectInfo {
  const javaRoot = findUpwards('pom.xml', cwd);
  const tsRoot =
    fs.existsSync(path.join(cwd, 'package.json')) ||
    fs.existsSync(path.join(cwd, 'tsconfig.json'));

  const root = javaRoot ?? cwd;
  const name = path.basename(root);

  if (javaRoot && tsRoot) return { lang: 'both', root, name };
  if (javaRoot) return { lang: 'java', root: javaRoot, name };
  if (tsRoot) return { lang: 'ts', root: cwd, name };
  return { lang: 'unknown', root: cwd, name };
}

function findUpwards(filename: string, from: string): string | null {
  let dir = from;
  for (let i = 0; i < 6; i++) {
    if (fs.existsSync(path.join(dir, filename))) return dir;
    const parent = path.dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  return null;
}

File: crates/reap/src/detect.rs
use std::path::{Path, PathBuf};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Lang {
    Java,
    Unknown,
}

pub struct ProjectInfo {
    pub lang: Lang,
    pub root: PathBuf,
    pub name: String,
}

pub fn detect_project(cwd: &Path) -> ProjectInfo {
    match find_upwards("pom.xml", cwd) {
        Some(root) => {
            let name = root
                .file_name()
                .map(|n| n.to_string_lossy().into_owned())
                .unwrap_or_default();
            ProjectInfo { lang: Lang::Java, root, name }
        }
        None => ProjectInfo { lang: Lang::Unknown, root: cwd.to_path_buf(), name: String::new() },
    }
}

fn find_upwards(filename: &str, from: &Path) -> Option<PathBuf> {
    let mut dir = from;
    for _ in 0..6 {
        if dir.join(filename).exists() {
            return Some(dir.to_path_buf());
        }
        match dir.parent() {
            Some(parent) if parent != dir => dir = parent,
            _ => break,
        }
    }
    None
}
